File: backend/app/matching/skill_matcher.py
from typing import List, Dict, Any
from rapidfuzz import fuzz
from app.matching.skill_normalizer import SkillNormalizer
from app.schemas.resume_schema import SkillItem


class SkillMatcher:
    @classmethod
    def match_skills(
        cls,
        candidate_skills: List[SkillItem],
        mandatory_jd_skills: List[str],
        preferred_jd_skills: List[str]
    ) -> Dict[str, Any]:
        matched = []
        partial = []
        missing = []

        candidate_map = {}
        for s in candidate_skills:
            norm_name = SkillNormalizer.normalize(s.name)
            candidate_map[norm_name.lower()] = s

        all_jd_skills = []
        for s in mandatory_jd_skills:
            all_jd_skills.append((s, True))
        for s in preferred_jd_skills:
            if s not in mandatory_jd_skills:
                all_jd_skills.append((s, False))

        for jd_skill, is_mandatory in all_jd_skills:
            norm_jd = SkillNormalizer.normalize(jd_skill)
            norm_jd_lower = norm_jd.lower()

            # 1. Exact / Normalized Match
            if norm_jd_lower in candidate_map:
                cand_skill = candidate_map[norm_jd_lower]
                matched.append({
                    "skill": jd_skill,
                    "normalized_skill": norm_jd,
                    "evidence": cand_skill.evidence or f"Found in resume as {cand_skill.name}",
                    "match_type": "EXACT",
                    "confidence": 1.0,
                    "is_mandatory": is_mandatory
                })
                continue

            # 2. Semantic / Related Match (Fuzzy ratio & substring containment)
            best_partial_match = None
            best_score = 0.0

            for cand_norm_lower, cand_skill in candidate_map.items():
                if SkillNormalizer.are_strictly_inequivalent(jd_skill, cand_skill.name):
                    continue

                # Substring inclusion (e.g. AWS EC2 -> AWS, Spring Boot -> Spring)
                if norm_jd_lower in cand_norm_lower or cand_norm_lower in norm_jd_lower:
                    score = 0.85
                else:
                    ratio = fuzz.ratio(norm_jd_lower, cand_norm_lower)
                    score = ratio / 100.0 if ratio >= 75 else 0.0

                if score > best_score and score >= 0.70:
                    best_score = score
                    best_partial_match = {
                        "skill": jd_skill,
                        "normalized_skill": norm_jd,
                        "candidate_skill": cand_skill.name,
                        "evidence": cand_skill.evidence or f"Related skill found: {cand_skill.name}",
                        "match_type": "PARTIAL",
                        "confidence": round(best_score, 2),
                        "is_mandatory": is_mandatory
                    }

            if best_partial_match:
                partial.append(best_partial_match)
            else:
                missing.append({
                    "skill": jd_skill,
                    "normalized_skill": norm_jd,
                    "evidence": "No evidence found in candidate resume",
                    "match_type": "MISSING",
                    "confidence": 0.0,
                    "is_mandatory": is_mandatory
                })

        # Calculate scores
        total_mandatory = len(mandatory_jd_skills)
        mandatory_matched_count = sum(1 for m in matched if m["is_mandatory"])
        mandatory_partial_count = sum(1 for p in partial if p["is_mandatory"])
        mandatory_score = ((mandatory_matched_count + (0.5 * mandatory_partial_count)) / total_mandatory * 100.0) if total_mandatory > 0 else 100.0

        total_preferred = len(preferred_jd_skills)
        pref_matched_count = sum(1 for m in matched if not m["is_mandatory"])
        pref_partial_count = sum(1 for p in partial if not p["is_mandatory"])
        preferred_score = ((pref_matched_count + (0.5 * pref_partial_count)) / total_preferred * 100.0) if total_preferred > 0 else 100.0

        return {
            "matched": matched,
            "partial": partial,
            "missing": missing,
            "mandatory_score": min(100.0, max(0.0, round(mandatory_score, 2))),
            "preferred_score": min(100.0, max(0.0, round(preferred_score, 2)))
        }

File: backend/src/matching/skillMatcher.js
const { normalizeSkill } = require('../ai/extractors');
const strictPairs = require('../../data/skill_aliases.json').strict_inequivalent || [];

function equivalent(a, b) {
  const left = normalizeSkill(a).toLowerCase();
  const right = normalizeSkill(b).toLowerCase();
  if (left === right || left.includes(right) || right.includes(left)) return true;
  if (strictPairs.some(pair => pair.map(item => normalizeSkill(item).toLowerCase()).includes(left) && pair.map(item => normalizeSkill(item).toLowerCase()).includes(right))) return false;
  return left.split(/[^a-z0-9]+/).some(token => token.length > 2 && right.includes(token));
}

function matchSkills(candidateSkills, mandatorySkills, preferredSkills) {
  const matched = [], partial = [], missing = [];
  const requirements = [...mandatorySkills.map(skill => [skill, true]), ...preferredSkills.filter(skill => !mandatorySkills.includes(skill)).map(skill => [skill, false])];
  for (const [skill, isMandatory] of requirements) {
    const exact = candidateSkills.find(candidate => normalizeSkill(candidate.name).toLowerCase() === normalizeSkill(skill).toLowerCase());
    if (exact) { matched.push({ skill, normalized_skill: normalizeSkill(skill), evidence: exact.evidence, match_type: 'EXACT', confidence: 1, is_mandatory: isMandatory }); continue; }
    const related = candidateSkills.find(candidate => equivalent(skill, candidate.name));
    if (related) partial.push({ skill, normalized_skill: normalizeSkill(skill), candidate_skill: related.name, evidence: related.evidence, match_type: 'PARTIAL', confidence: 0.85, is_mandatory: isMandatory });
    else missing.push({ skill, normalized_skill: normalizeSkill(skill), evidence: 'No evidence found in candidate resume', match_type: 'MISSING', confidence: 0, is_mandatory: isMandatory });
  }
  const score = (items, total) => total ? ((items.matched + items.partial * 0.5) / total) * 100 : 100;
  return { matched, partial, missing, mandatory_score: Math.min(100, Number(score({ matched: matched.filter(item => item.is_mandatory).length, partial: partial.filter(item => item.is_mandatory).length }, mandatorySkills.length).toFixed(2))), preferred_score: Math.min(100, Number(score({ matched: matched.filter(item => !item.is_mandatory).length, partial: partial.filter(item => !item.is_mandatory).length }, preferredSkills.length).toFixed(2))) };
}

module.exports = { matchSkills };
